lib/galaxy/tool_util/verify/interactor.py

```diff
@@ -715,16 +709,35 @@ def get_element(elements, id):
             raise AssertionError(message)
 
     def verify_elements(element_objects, element_tests):
+        sorted_test_ids = [None] * len(element_tests)
         for element_identifier, element_test in element_tests.items():
             if isinstance(element_test, dict):
                 element_outfile, element_attrib = None, element_test
             else:
                 element_outfile, element_attrib = element_test
+            sorted_test_ids[element_attrib["element_index"]] = element_identifier
+
+        i = 0
+        for element_identifier in sorted_test_ids:
+            element_test = element_tests[element_identifier]
+            if isinstance(element_test, dict):
+                element_outfile, element_attrib = None, element_test
+            else:
+                element_outfile, element_attrib = element_test
+
+            element = None
+            while i < len(element_objects):
+                if element_objects[i]["element_identifier"] == element_identifier:
+                    element = element_objects[i]
+                    i += 1
+                    break
+                i += 1
 
-            element = get_element(element_objects, element_identifier)
-            if not element:
-                template = "Failed to find identifier [%s] for testing, tool generated collection elements [%s]"
-                message = template % (element_identifier, element_objects)
+            if element is None:
+                template = "Failed to find identifier '%s' of test collection %s in the tool generated collection elements %s (at the correct position)"
```

Comment:
Isn't `the corrent position` `i` ?

Reply:
and wouldn't this mean you can simplify the logic to

```python
for element_index, expected_element_identifier in sorted_test_ids:
    assert element_objects[element_index]["element_identifier"] == expected_element_identifier
```
?

(You could add some length handling for missing elements)

Reply:
No, the test is just checking if the elements listed in the test occur in the same order in the generated elements. 

Thereby the tool author does not need to list all elements (which might be very many).

Reply:
Alright, then how about this:
```
for element_index, expected_element_identifier in enumerate(sorted_test_ids):
    if expected_element_identifier is None:
        break
    assert element_objects[element_index]["element_identifier"] == expected_element_identifier
```
That way it's clear at which position the test data differs, and I think the logic is a bit easier to follow ?

Reply:
I don't see why this should work :( .. 

The `None` entries from here: https://github.com/galaxyproject/galaxy/blob/a5b898c35bee08ada119bd0f7c3912242b1bae1c/lib/galaxy/tool_util/verify/interactor.py#L728

should all be overwritten by the actual identifiers. I only use `[None] * len(element_tests)` to initialize a list of the needed length.

I extended the test in order to clarify what I meant. The tool new generates 10 elements, but the test only uses three of them to verify the sorting order. The number of generated elements is tested with the `count` attribute. 

Formally spoken we test if the list of elements in the test is a subsequence of the generated elements.

Reply:
what is element_tests ? Is that not the list of specified tests?

Reply:
Its a `dict`, looking like this

```
{'1': [None, {'assert_list': [{'attributes': {'expression': '^.*$'}, 'children': [], 'tag': 'has_text_matching'}], 'checksum': None, 'compare': 'diff', 'decompress': False, 'delta': 10000, 'delta_frac': None, 'element_index': 0, 'elements': {}, 'extra_files': [], 'lines_diff': 0, 'md5': None, 'metadata': {}, 'primary_datasets': {}, 'sort': False}], '10': [None, {'assert_list': [{'attributes': {'expression': '^.*$'}, 'children': [], 'tag': 'has_text_matching'}], 'checksum': None, 'compare': 'diff', 'decompress': False, 'delta': 10000, 'delta_frac': None, 'element_index': 2, 'elements': {}, 'extra_files': [], 'lines_diff': 0, 'md5': None, 'metadata': {}, 'primary_datasets': {}, 'sort': False}], '2': [None, {'assert_list': [{'attributes': {'expression': '^.*$'}, 'children': [], 'tag': 'has_text_matching'}], 'checksum': None, 'compare': 'diff', 'decompress': False, 'delta': 10000, 'delta_frac': None, 'element_index': 1, 'elements': {}, 'extra_files': [], 'lines_diff': 0, 'md5': None, 'metadata': {}, 'primary_datasets': {}, 'sort': False}]}
```

Reply:
Maybe calling it element_index is a bad idea if it isn't expected to match the actual element index. Something like expected_order_index maybe would be better?

I don't really have the throughput to deal with this, but maybe this change requiring the order to match should be hidden behind a profile version tag for the tool? If there wasn't a ton of tool breakage maybe it is not worth the effort, I will just update Planemo regardless.